packages/cli/command_help.pony

```diff
@@ -0,0 +1,196 @@
+    w.write("\n")
+    w.write(spec.descr + "\n")
+    w.write("\n")
+
+    let options = all_options()
+    if options.size() > 0 then
+      w.write("Options:\n")
+      print_options(w, options)
+    end
+    if spec.commands.size() > 0 then
+      w.write("\nCommands:\n")
+      print_commands(w)
+    end
+    let args = spec.args
+    if args.size() > 0 then
+      w.write("\nArgs:\n")
+      print_args(w, args)
+    end
+
+  fun print_usage(w: Writer) =>
+    w.write("usage: " + fullname())
+    if any_options() then
+      w.write(" [<options>]")
+    end
+    if spec.commands.size() > 0 then
+      w.write(" <command>")
+    end
+    if spec.args.size() > 0 then
+      for a in spec.args.values() do
+        w.write(" " + a.help_string())
+      end
+    else
+      w.write(" [<args> ...]")
+    end
+    w.write("\n")
+
+  fun print_options(w: Writer, options: Array[OptionSpec box] box) =>
+    let cols = Array[(String,String)]()
+    for o in options.values() do
+      cols.push(("  " + o.help_string(), o.descr))
+    end
+    Columns.print(w, cols)
+
+  fun print_commands(w: Writer) =>
+    let cols = Array[(String,String)]()
+    _list_commands(spec, cols, 1)
+    Columns.print(w, cols)
+
+  fun _list_commands(cs: CommandSpec box, cols: Array[(String,String)], level: USize) =>
+    for c in cs.commands.values() do
+      cols.push((Columns.indent(level*2) + c.help_string(), c.descr))
+      _list_commands(c, cols, level + 1)
+    end
+
+  fun print_args(w: Writer, args: Array[ArgSpec] box) =>
+    let cols = Array[(String,String)]()
+    for a in args.values() do
+      cols.push(("  " + a.help_string(), a.descr))
+    end
+    Columns.print(w, cols)
+
+  fun any_options(): Bool =>
+    if spec.options.size() > 0 then
+      true
+    else
+      match parent
+      | let p: CommandHelp => p.any_options()
+      else
+        false
+      end
+    end
+
+  fun all_options(): Array[OptionSpec box] =>
+    let options = Array[OptionSpec box]()
+    _all_options(options)
+    options
+
+  fun _all_options(options: Array[OptionSpec box]) =>
+    match parent
+    | let p: CommandHelp => p._all_options(options)
+    end
+    for o in spec.options.values() do
+      options.push(o)
+    end
+
+"""
+This interface and two classes allow the help output to go to a general Writer
+that can be implemented for string creation or to an OutStream. Or other
+targets. This could be polished up and moved into a more central package.
+"""
+interface Writer
+  fun ref write(data: ByteSeq)
+
+class StringWriter
+  let _s: String ref = String
+
+  fun ref write(data: ByteSeq) =>
+    _s.append(data)
+
+  fun string(): String iso^ =>
+    _s.clone()
+
+class OutWriter
+  let _os: OutStream tag
+
+  new create(os: OutStream tag) =>
+    _os = os
+
+  fun ref write(data: ByteSeq) =>
+    _os.write(data)
+
+
+primitive Columns
+  fun indent(n: USize): String =>
+    let spaces = "                                "
+    spaces.substring(0, n.isize())
```

Comment:
Possibly. This output support is intended as a temporary solution until we can add something more complete and polished.